test/test_integration.py

```diff
@@ -324,3 +327,140 @@ def test_gallery(self):
+        def update_location_on_page(hash_val):
+            if hash_val is None:
+                return ''
+
+            return hash_val
+
+        @app.callback(
+            output=Output(component_id='test-search', component_property='children'),
+            inputs=[Input(component_id='test-location', component_property='search')])
+        def update_location_on_page(search):
+            if search is None:
+                return ''
+
+            return search
+
+        @app.callback(
+            output=Output(component_id='test-location', component_property='pathname'),
+            inputs=[Input(component_id='test-button', component_property='n_clicks')],
+            state=[State(component_id='test-location', component_property='pathname')])
+        def update_pathname(n_clicks, current_pathname):
+            if n_clicks is not None:
+                return '/new/pathname'
+
+            return current_pathname
+
+        self.startServer(app=app)
+
+        def _waiter():
+            time.sleep(1)
+            try:
+                self.wait_for_element_by_id('waitfor')
+            except Exception as e:
+                print(self.wait_for_element_by_id(
+                    '_dash-app-content').get_attribute('innerHTML'))
+                raise e
+
+        self.snapshot('link -- location')
+
+        # Check that link updates pathname
+        self.driver.find_element_by_id('test-link').click()
+
+        self.snapshot('link --- /test/pathname')
+        self.assertEqual(
+            self.driver.current_url.replace('http://localhost:8050', ''),
+            '/test/pathname')
+        self.assertEqual(
+            self.driver.find_element_by_id('test-pathname').text,
+            '/test/pathname')
+
+        # Check that hash is updated in the Location
+        self.driver.find_element_by_id('test-link-hash').click()
+
+        self.snapshot('link -- /test/pathname#test')
+        self.assertEqual(self.driver.find_element_by_id('test-pathname').text, '/test/pathname')
+        self.assertEqual(self.driver.find_element_by_id('test-hash').text, '#test')
+
+        # Check that search is updated in the Location -- note that this goes through href and therefore wipes the hash
+        self.driver.find_element_by_id('test-link-search').click()
+
+        self.snapshot('link -- /test/pathname?testQuery=testValue')
+        self.assertEqual(self.driver.find_element_by_id('test-search').text, '?testQuery=testValue')
+        self.assertEqual(self.driver.find_element_by_id('test-hash').text, '')
+
+        # Check that pathname is updated through a Button click via props
+        self.driver.find_element_by_id('test-button').click()
+        time.sleep(1)  # Need to wait for the callback to fire TODO is there a better way to wait?
+
+        self.snapshot('link -- /new/pathname?testQuery=testValue')
+        self.assertEqual(self.driver.find_element_by_id('test-pathname').text, '/new/pathname')
```

Comment:
for example, you _should_ be able to remove the `time.sleep` and replace it with a `wait_for` here and place the snapshot below the assertions